sssd_test_framework/utils/pam.py

```diff
@@ -0,0 +1,284 @@
+            raise ValueError("No data!")
+
+        match = self.util.host.ssh.run(f"augtool {self.args} match {self.path}/access").stdout.split("\n")
+
+        self.util.logger.info(f"Deleting node in Augeas tree on {self.util.host.hostname}")
+        for i in range(1, 1 + len(match)):
+            for k, v in value.items():
+                if k and v in match[i]:
+                    self.util.host.ssh.run(f"augtool {self.args} --autosave rm {self.path}/access[{i}]")
+
+    def config_set(self, value: list[dict[str, str]]) -> None:
+        """
+        Configure access configuration file.
+        :param value: Access rule
+        :type value: list[list[str]]
+        :return: None
+        """
+        if value is None:
+            raise ValueError("No data!")
+
+        count = 1
+        for i in value:
+            self.cmd = self.cmd + f"set {self.path}/access[{count}] " + i["access"] + "\n"
+            self.cmd = self.cmd + f"set {self.path}/access[{count}]/user " + i["user"] + "\n"
+            self.cmd = self.cmd + f"set {self.path}/access[{count}]/origin " + i["origin"] + "\n"
+            count = count + 1
+
+
+class PAMFaillock:
+    """
+    Management of PAM Faillock on the client host.
+
+    .. code-block:: python
+        :caption: Example usage
+
+        @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
+        def test_example(client: Client, provider: GenericProvider):
+            # Add user
+            provider.user("user-1").add()
+
+            # Setup faillock
+            faillock = client.pam.faillock()
+            faillock.config_set({"deny": "3", "unlock_time": "300"})
+            faillock.config_write()
+
+            client.sssd.common.pam(["with-faillock"])
+
+            # Start SSSD
+            client.sssd.start()
+
+            # Check the results
+            assert client.auth.ssh.password("user-1", "Secret123")
+
+            # Three failed login attempts
+            for i in range(3):
+                assert not client.auth.ssh.password("user-1", "bad_password")
+
+            assert not client.auth.ssh.password("user-1", "Secret123")
+
+            # Reset user lockout
+            client.pam.faillock("user-1").reset
+            assert client.auth.ssh.password("user-1", "Secret123")
+    """
+
+    def __init__(self, util: PAMUtils, file: str = "/etc/security/faillock.conf") -> None:
+        """
+        :param util: PAMUtils object
+        :type util: PAMUtils
+        :param file: Faillock configuration file
+        :type file: str, optional
+        """
+        self.util: PAMUtils = util
+        self.file: str = file
+        self.path: str = "/files" + self.file
+        self.args: str = f'--noautoload --transform "Simplevars.lns incl {self.file}"'
+        self.cmd: str = ""
+
+    def config_write(self) -> None:
+        """
+        Write faillock configuration.
+        :return: Self
+        :rtype: PAMFaillock
+        """
+        self.util.fs.backup(self.file)
```

Comment:
Same as `PAMAccess` regarding setup and teardown.

Reply:
I thought that util.fs.backup() automatically restored the file after the test execution, so by using this we didn't need to write additional setup and teardown functions for files?

Reply:
You are right, sorry for the noise.